db/syncPostsToGrapher.ts

```diff
@@ -9,7 +9,101 @@ import { postsTable, select } from "./model/Post.js"
+        "select ID, post_content from wp_posts where post_type='wp_block' AND post_status = 'publish'"
+    )
+
+    const allBlocks = keyBy(blocks, "ID")
+    return allBlocks
+}
+
+type ReplacerFunction = (
+    _match: string,
+    _firstPattern: string,
+    _offset: number,
+    fullString: string,
+    matches: Record<string, string>
+) => string
+
+/** Function that takes an object where the keys are reusable block ids and the values are
+    ReusableBlocks and returns a replacer function that can be used as a replacer when using
+    string.replace(regex, replacer_function) to replace matches that capture the id with
+    the content of the block from the blocks input param with the same id as the key */
+function buildReplacerFunction(
+    blocks: Record<string, ReusableBlock>
+): ReplacerFunction {
+    return (
+        _match: string,
+        _firstPattern: string,
+        _offset: number,
+        fullString: string,
+        matches: Record<string, string>
+    ) => {
+        const block = blocks[matches["id"].toString()]
+        if (block) return block.post_content
+        else return fullString
+    }
+}
+
+function replaceReusableBlocksRecursive(
+    content: string,
+    replacerFunction: ReplacerFunction
+): string {
+    // Resolve references by using replace with a regex and a function.
+    // There are two special cases - one is that not all refs resolve (🤨)
+    // in which case we leave the original ref comment as is; the second is
+    // that some blocks reference other blocks (🎉 recursion!) and so we
+    // check in a loop if the regexp still matches and run replace again.
+    // Because not all refs resolve we have to limit the number of attempts - for now
+    // we try it 3 times which should be plenty for all reasonably sane scenarios
+    let contentWithBlocksInlined = content
+    for (
+        let recursionLevelsRemaining = 3;
+        recursionLevelsRemaining > 0 &&
+        contentWithBlocksInlined.match(blockRefRegex);
+        recursionLevelsRemaining--
+    ) {
+        contentWithBlocksInlined = contentWithBlocksInlined.replace(
+            blockRefRegex,
+            replacerFunction
+        )
+    }
+    return contentWithBlocksInlined
+}
+
+type BlockResolveFunction = (content: string) => string
+
+/** This function fetches all reusable blocks and then returns a function that
+    takes a post content and returns the content with all references to blocks resolved.
+    To do it's work this function uses a database connection to fetch all blocks when it is called
+    that has to be awaited but then the function that is returned is then a simple lookup implementation.
+    This was implemented as a closure for nicer re-use and encapsulation.
+
+    @example
+    const content = "some content with a <!-- wp:block {\"ref\":123} /--> reference"
+    const replacerFn = await buildReusableBlocksResolver()
+    const dereferencedConentet = replacerFn(content)
+     */
```

Comment:
What do others think about JSDoc comments? I like them for functions that are doing something a bit more complex so that you can just hover over them at the site of usage and see those comments in the on-hover window

Reply:
yep I agree